Use the same settings UI as the devtools perf panel

past opened the pull request:
This PR copies over the new devtools performance panel settings style to the add-on. I literally copied over the generated markup and tweaked the JS event handling and initialization to match the simpler add-on architecture. I tried to keep things sufficiently generic, so there is handling for the 'java' profiler feature, even though the add-on doesn't currently work on Fennec (#33). 

I also took the liberty to remove the Information widget, which I believe we no longer consider useful. I inlined some colors that were using CSS variables in the devtools panel, since we don't yet support themes in the add-on.

gregtatum commented:
I didn't get to this today, I'll try and look at it on Monday.

gregtatum commented:
UI feedback:

* I don't think we should have 2 levels of collapsed information. In DevTools the settings are open by default, but the threads and features are collapsed. So I think we should either remove the first or second level of hiding.

* The "Apply" button drops off the screen when expanding things fully open, which I think it should always be visible. I think a viable solution would be to anchor the button at the bottom, and a nice (optional) UX feature would be to only show the apply button once a setting is changed. <img width="397" alt="image" src="https://user-images.githubusercontent.com/1588648/45959144-d0b57d00-bfde-11e8-8113-6d3823e5d5c3.png">

* The width feels a little cramped, especially with the two column feature labels. I feel like we should expand it to something around 400-450px.

* Expanding the features or threads menu causes an awkward animation transition where the popup window expands in size, but the inner contents do not until a second later: <img width="372" alt="image" src="https://user-images.githubusercontent.com/1588648/45959663-0c9d1200-bfe0-11e8-84b6-eb387991479b.png">

* The threads and features showing and hiding feature is causing my computer to play the error ding. I'm not sure why, and it feels a bit intermittent.

In summary, my recommendation would be to remove the second level indent; to have the threads and features non-collapsible and always open. Also when I was playing with this idea locally, there is some styling that is making the spacing between the dropdown labels inconsistent. Note that this issue exists in the DevTools UI, but feels more blatant in the tighter UI space of the pop-up.

gregtatum commented on popup.css:
```diff
@@ -255,20 +255,6 @@ kbd {
   background-color: hsla(0, 90%, 40%, 0.5);
 }
 
-.relevancy-level {
```
@mstange Are you OK with the relevancy level going away? It seems fine to me. I don't really recall the decision to do this in the DevTools panel.

<img width="398" alt="image" src="https://user-images.githubusercontent.com/1588648/45960348-ab763e00-bfe1-11e8-8898-df0fc037d482.png">

mstange replied:
Yes that seems fine. It was a cute but rather useless gimmick.

gregtatum commented on background.js:
```diff
@@ -164,27 +155,39 @@ async function restartProfiler() {
-    let platform = await browser.runtime.getPlatformInfo();
-    if (platform.os === 'mac') {
-      enableScreenshots = true;
+    const features = {
```
This is already a fairly large UI change, which I'm more concerned about than in other projects due to the lack of any testing infrastructure in place on the add-on. This is feature change feels out of scope, and will make regression tracking harder. Would you mind breaking it out into a separate PR?

past replied:
I'm not sure I understand which part of these changes you find concerning. The bulk of the changes in this file is for the added checkboxes that require initializing of their corresponding feature controls, so that's probably not what you meant. 

Another change is moving the 'leaf' feature initialization along with the rest, because currently it's very easy to miss and assume that it is uninitialized. That's just 2 lines of code, so again, probably not what you meant.

The last change is about making the "java" feature conditional on the platform instead of always disabled by default. I admit that this is a bit egregious, given that the add-on doesn't support Fennec at the moment, but it seems simple enough, generalizes the platform-dependent initialization code, and gives us less work for #33. Is this part what you'd like me to move to a separate PR?

gregtatum replied:
Coming back to this, I think I overlooked that the following snippet was in the original code.

```
if (platform.os === 'mac') {
      enableScreenshots = true;
}
```

There is still some behavior change here, but it's less than I thought. I'd be fine if you kept it in this PR.

> I'm not sure I understand which part of these changes you find concerning

Also, some of my comments were on the state of the codebase, not specifically on your code changes.

gregtatum commented on popup.js:
```diff
@@ -204,38 +203,85 @@ document
  * will adjust the profiler state when changed.
  */
-async function setupFeatureCheckbox(featureName) {
+async function setupFeatureCheckbox(name) {
```
This first part seems like it should be broken out with the other "feature" code.

past replied:
Which "first part" are you referring to? The rename of `featureName` to `name` or some of the code earlier in this file?

If you mean the argument renaming, I can certainly take it out, but it seemed a bit more uniform with `setupThreadCheckbox(name)`, even though I didn't end up combining them as I was initially planning.

gregtatum replied:
I was certainly not communicating well on this PR. I believe I was referring to the:

```

  // Java profiling is only meaningful on android.
  if (name == 'java') {
    if (platform.os !== 'android') {
      document.querySelector('#java').style.display = 'none';
      return;
    }
  } 
```

I changed my thoughts on it, please keep it.

gregtatum commented on popup.js:
```diff
@@ -204,38 +203,85 @@ document
+  );
+}
+
+for (let name of features) {
```
Optional: I'm a little concerned with imperative code being run intermixed with function definitions. However, there are many query selectors around doing something similar around this code. These `for` loops stand out as a bit more egregious. Typically I prefer a setup/init function to make the order of operations a bit more explicit and clear. Really though, cleaning this up I think would require a separate PR.

gregtatum commented on popup.js:
```diff
@@ -34,15 +63,18 @@ function renderControls(state) {
-  document.querySelector('.tasktracer-checkbox').value = state.tasktracer;
-  document.querySelector('.threads-textbox').value = state.threads;
+
+  for (let name of features) {
```
Optional: My preference is to `const` everything that can be, although we have no linting here to enforce it. In this for loop `name` is never rebound.

```js
for (const name of features) { ... }
```

past commented:
I believe I have addressed all of your feedback, with the exception of the two optional change requests (displaying the 'apply' button only after changes, and moving popup initialization to a function). I'll try to look into those in followups as time permits.

canaltinova commented:
I can take this over since I also need to write a patch that depends on this(for the fixed-time-window profiling mode).

canaltinova commented:
Addressed the review comments. Now features side looks like this:

<img width="357" alt="screen shot 2018-10-22 at 6 54 23 pm" src="https://user-images.githubusercontent.com/466239/47310930-bea81800-d638-11e8-880f-01fd78aea9fb.png">

@julienw Could you look at the patch and also check if the checkbox and the description alignment is also like this on your side?

julienw commented:
There's still a problem with the feature labels, see the screenshot:
![image](https://user-images.githubusercontent.com/454175/47352601-86491e00-d6ba-11e8-88c9-63c16108c7ee.png)

The problem is the `width` in https://github.com/devtools-html/Gecko-Profiler-Addon/pull/125/files#diff-b5dfc3ff7c7bccd1bec16cd3bc5f663eR388 that's now useless:
```css
.perf-settings-feature-name {
  width: 100px; // <-- you can remove this
  color: #0060df;
  line-height: 1.6;
}
```

julienw commented on popup.css:
```diff
@@ -12,7 +12,7 @@ html {
 body {
   margin: 0;
-  width: 400px;
+  max-width: 32em;
```
ah, and don't forget to file a bug in Firefox about that.
Then add a comment explaining that if we use `width` instead of `max-width` we have an horizontal screenshot in Firefox, with the bug #.
